godot-core/src/builtin/array.rs

```diff
@@ -126,13 +111,114 @@ impl Array {
+    pub fn sort_unstable(&mut self) {
+        self.as_inner().sort();
+    }
+
+    /// Shuffles the array such that the items will have a random order. This method uses the
+    /// global random number generator common to methods such as `randi`. Call `randomize` to
+    /// ensure that a new seed will be used each time if you want non-reproducible shuffling.
+    pub fn shuffle(&mut self) {
+        self.as_inner().shuffle();
+    }
+
+    /// Asserts that the given index refers to an existing element.
+    ///
+    /// # Panics
+    ///
+    /// If `index` is out of bounds.
+    fn check_bounds(&self, index: usize) {
+        let len = self.len();
+        assert!(
+            index < len,
+            "Array index {index} is out of bounds: length is {len}",
+        );
+    }
+
+    /// Returns a pointer to the element at the given index.
+    ///
+    /// # Panics
+    ///
+    /// If `index` is out of bounds.
+    fn ptr(&self, index: usize) -> *const Variant {
+        self.check_bounds(index);
+        // SAFETY: We just checked that the index is not out of bounds.
+        let ptr = unsafe { self.ptr_unchecked(index) };
+        assert!(!ptr.is_null());
+        ptr
+    }
+
+    /// Returns a mutable pointer to the element at the given index.
+    ///
+    /// # Panics
+    ///
+    /// If `index` is out of bounds.
+    fn ptr_mut(&self, index: usize) -> *mut Variant {
+        self.check_bounds(index);
+        // SAFETY: We just checked that the index is not out of bounds.
+        let ptr = unsafe { self.ptr_mut_unchecked(index) };
+        assert!(!ptr.is_null());
+        ptr
+    }
+
+    /// Returns a pointer to the element at the given index.
+    ///
+    /// # Safety
+    ///
+    /// Calling this with an out-of-bounds index is undefined behavior.
+    unsafe fn ptr_unchecked(&self, index: usize) -> *const Variant {
+        let item_ptr: sys::GDExtensionVariantPtr =
+            (interface_fn!(array_operator_index_const))(self.sys(), to_i64(index));
+        item_ptr as *const Variant
+    }
+
+    /// Returns a mutable pointer to the element at the given index.
+    ///
+    /// # Safety
+    ///
+    /// Calling this with an out-of-bounds index is undefined behavior.
+    unsafe fn ptr_mut_unchecked(&self, index: usize) -> *mut Variant {
+        let item_ptr: sys::GDExtensionVariantPtr =
+            (interface_fn!(array_operator_index))(self.sys(), to_i64(index));
+        item_ptr as *mut Variant
+    }
+
+    #[doc(hidden)]
+    pub fn as_inner(&self) -> inner::InnerArray {
+        inner::InnerArray::from_sys_ptr(self.sys())
+    }
+
+    /// Changes the generic type on this array, without changing its contents. Needed for API
+    /// functions that return a variant array even though we know its type.
+    fn assume_type<U: VariantMetadata>(self) -> TypedArray<U> {
+        // SAFETY: The memory layout of `TypedArray<T>` does not depend on `T`.
+        unsafe { std::mem::transmute(self) }
+    }
```

Comment:
This should probably call `check_type()`. 

Even if the layout is the same, the entire invariant of the class is that each element has type `U`.
It isn't unimaginable that future code relies on this, without thinking of this (unchecked yet safe) boundary.

Reply:
Note that `assume_typed` is not public. I think we should have a public wrapper like `to_typed()` that _does_ check the type (and returns `Result`).

Reply:
Public/private is generally orthogonal to safe/`unsafe`. This method can be used to subvert the type system in a way that can cause UB. Not in the current implementation (as far as I can see at least), but as soon as any code starts relying on the fact that a `Array<T>` holds indeed variants of type `T`. Which is a generally reasonable assumption*.

Imagine we optimize one of the accesses in the future:
```rs
pub fn get(&self, index: usize) -> T {
    ...
    T::from_variant(variant)
}
```
becomes
```rs
pub fn get(&self, index: usize) -> T {
    ...
    // SAFETY: every Variant in Array<T> has T
    unsafe { T::from_variant_unchecked(variant) }
}
```
Now you can argue that this `unsafe` block introduced UB. But it's not true -- its `// SAFETY` assumption relies on the invariant of the type `Array<T>`. The function that allows this invariant to be violated in the first place is `assume_typed()`, if used wrongly. 

So there are two options:
* `assume_typed()` must be `unsafe`, as its _external_ invariant is that it is always called on arrays with the correct type
* `assume_typed()` verifies that invariant and is thus safe.

---

\* As you brought up, there are edge cases where "`Array<T>` always holds `T` variants" is hard to enforce, but I'd like to try. Even if this turned out to be wrong, I wouldn't see it as a reason to open more doors that can violate invariants.

Reply:
Fair point. In an ideal world, I would mark `assume_typed()` as `unsafe` because all current call sites actually do uphold the invariant. However, we don't know what curve balls Godot's typing will throw us in the future, so an extra `check_type()` is the safer route.

Reply:
Turns out, we actually need to break the invariant sometimes. `InnerArray::append_array()` for example takes any `Array`, not necessarily a `TypedArray` of the same type as `self`. So I'm going to mark this `unsafe` instead.

Reply:
But why does that make it `unsafe`? The method `assume_typed()` would just panic on `check_type()`, no?

In other words, it's a logic error if we don't allow to append `Array<String>` to `Array<Variant>`, but not a memory safety issue.

Reply:
We can't call `check_type()` inside `assume_typed()` because it would panic for a legitimate, but internal, use case: conversion from `TypedArray<T>` to `TypedArray<Variant>` in order to pass it to `InnerArray::append_array()`. You mentioned two options above; we can't use the second, so I went with the first.